### PR TITLE
AEA-219  Identify the packages by unique_id

### PR DESCRIPTION
**Pull request by Totoual**
## Proposed changes

Generate a uuid for each connection/protocol/skill so it can be unique.

## Fixes

Fixes: #513 and #155 

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../master/CONTRIBUTING.rst) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: aea/configurations/base.py [hunk 1]
```diff
@@ -197,6 +198,7 @@ def __init__(self,
         self.authors = authors
         self.version = version
         self.license = license
+        self.uuid = generate_uuid(author=self.authors, package_name=self.name, version=self.version)
```
Comment (DavidMinarsch) [inline aea/configurations/base.py]: @marcofavorito and @Totoual
- are we happy calling it uuid? It makes sense, but maybe not quite in line with crypto lingo.
- we should allow for a fourth, optional field in the `generate_uuid` function called `nonce`. Then we can by default set the nonce to None. If the user wants to use two packages with different configs then they can achieve that by feeding a `nonce` to the `generate_uuid` function. Note, we do not allow for the submission of the same package with two different nonces to the package registry, hence we do not need to keep track of the nonce in the config (it is only for local usage).
- I think we need to add a `user_name` field and replace the authors with the `user_name`
- All packages should have `user_name = fetchai`
- All packages should have their uuid in the config (only do this once we have finished everything else int his PR)
- Once this PR is finished, we need another PR which applies the usage of uuid to how packages are identified within the AEA framework, but we cannot do this for 2.0, hence please open an issue.

File: aea/helpers/base.py [hunk 2]
```diff
@@ -68,3 +70,13 @@ def locate(path):
...
+
+
+def generate_public_id(user_name: str, package_name: str, version: str, nonce: Optional[int] = None) -> str:
+    """Generate a unique id for the package."""
+    import hashlib
+    string_for_hash = "".join([user_name, package_name, version, str(nonce)])
+    m_hash = hashlib.sha3_256()
+    m_hash.update(string_for_hash.encode())
+    encoded_str = m_hash.digest().hex()
+    return encoded_str
```
Comment (panasevychol) [inline aea/helpers/base.py]: So what do we have on the output? Hashed string?

File: aea/configurations/base.py [hunk 3]
```diff
@@ -182,7 +183,7 @@ class ConnectionConfig(Configuration):
 
     def __init__(self,
                  name: str = "",
-                 authors: str = "",
+                 user_name: str = "",
```
Comment (panasevychol) [inline aea/configurations/base.py]: I would prefer `owner` or `owner_name` or smth. like that. To explain which user_name is that.

File: aea/helpers/base.py (hunk 2 above)
Comment (Totoual) [reply aea/helpers/base.py]: yeap

File: aea/configurations/base.py (hunk 3 above)
Comment (Totoual) [reply aea/configurations/base.py]: We will switch back to author to avoid any conflict.

File: aea/connections/scaffold/connection.yaml [hunk 4]
```diff
@@ -1,7 +1,8 @@
 name: scaffold_connection
```
Comment (DavidMinarsch) [inline aea/connections/scaffold/connection.yaml]: This should be called `scaffold` only

File: aea/protocols/scaffold/protocol.yaml [hunk 5]
```diff
@@ -1,6 +1,7 @@
 name: 'scaffold_protocol'
```
Comment (DavidMinarsch) [inline aea/protocols/scaffold/protocol.yaml]: This should be `scaffold` only

File: aea/skills/scaffold/skill.yaml [hunk 6]
```diff
@@ -1,7 +1,8 @@
 name: scaffold_skill
```
Comment (DavidMinarsch) [inline aea/skills/scaffold/skill.yaml]: This should be `scaffold` only

File: tests/data/aea-config.example.yaml [hunk 7]
```diff
@@ -1,6 +1,6 @@
 aea_version: 0.1.1
 agent_name: myagent
-authors: Fetch.AI Limited
```
Comment (DavidMinarsch) [inline tests/data/aea-config.example.yaml]: missing fingerprint field

File: tests/data/aea-config.example_w_keys.yaml [hunk 8]
```diff
@@ -1,6 +1,6 @@
 aea_version: 0.1.1
 agent_name: myagent
-authors: Fetch.AI Limited
```
Comment (DavidMinarsch) [inline tests/data/aea-config.example_w_keys.yaml]: missing fingerprint field

File: tests/data/dummy_aea/aea-config.yaml [hunk 9]
```diff
@@ -1,6 +1,6 @@
 aea_version: 0.1.7
 agent_name: Agent0
-authors: Fetch.AI Limited
```
Comment (DavidMinarsch) [inline tests/data/dummy_aea/aea-config.yaml]: missing fingerprint field

File: aea/helpers/base.py [hunk 10]
```diff
@@ -68,3 +70,13 @@ def locate(path):
         except AttributeError:
             return None
     return object
+
+
+def generate_fingerprint(author: str, package_name: str, version: str, nonce: Optional[int] = None) -> str:
+    """Generate a unique id for the package."""
+    import hashlib
+    string_for_hash = "".join([author, package_name, version, str(nonce)])
```
Comment (DavidMinarsch) [inline aea/helpers/base.py]: If the nonce = None, we should not convert it to the string, rather just have the first three

File: tests/data/dummy_aea/aea-config.yaml (hunk 9 above)
Comment (Totoual) [reply tests/data/dummy_aea/aea-config.yaml]: I don't think that we want fingerprint here.

File: tests/test_packages/test_skills/data/connection.yaml [hunk 11]
```diff
@@ -1,5 +1,5 @@
 name: gym
-authors: Fetch.AI Limited
+author: fetchai
 version: 0.1.0
```
Comment (DavidMinarsch) [inline tests/test_packages/test_skills/data/connection.yaml]: add fingerprint

**Pull request merged**